c/tskit/tables.c

```diff
@@ -8106,6 +8106,285 @@ tsk_table_collection_subset(
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    memset(X_node_map, 0xff, other->nodes.num_rows * sizeof(*X_node_map));
+    memset(individual_map, 0xff, other->individuals.num_rows * sizeof(*individual_map));
+    memset(population_map, 0xff, other->populations.num_rows * sizeof(*population_map));
+    memset(site_map, 0xff, other->sites.num_rows * sizeof(*site_map));
+
+    // checks input: node_mapping consistency and equivalency of tables on shared nodes
+    ret = tsk_union_init(self, other, node_mapping, num_nodes, options);
+    if (ret != 0) {
+        goto out;
+    }
+
+    // nodes, individuals, populations
+    for (k = 0; k < (tsk_id_t) num_nodes; k++) {
+        if (node_mapping[k] != TSK_NULL) {
+            X_node_map[k] = node_mapping[k];
+        } else {
+            ret = tsk_node_table_get_row(&other->nodes, k, &node);
+            if (ret < 0) {
+                goto out;
+            }
+            new_ind = TSK_NULL;
+            if (node.individual != TSK_NULL) {
+                if (individual_map[node.individual] == TSK_NULL) {
+                    tsk_individual_table_get_row(
+                        &other->individuals, node.individual, &ind);
+                    ret = tsk_individual_table_add_row(&self->individuals, ind.flags,
+                        ind.location, ind.location_length, ind.metadata,
+                        ind.metadata_length);
+                    if (ret < 0) {
+                        goto out;
+                    }
+                    individual_map[node.individual] = ret;
+                }
+                new_ind = individual_map[node.individual];
+            }
+            new_pop = TSK_NULL;
+            if (node.population != TSK_NULL) {
+                // keep same pops if add_populations is False
+                if (!add_populations) {
+                    population_map[node.population] = node.population;
+                }
+                if (population_map[node.population] == TSK_NULL) {
+                    tsk_population_table_get_row(
+                        &other->populations, node.population, &pop);
+                    ret = tsk_population_table_add_row(
+                        &self->populations, pop.metadata, pop.metadata_length);
+                    if (ret < 0) {
+                        goto out;
+                    }
+                    population_map[node.population] = ret;
+                }
+                new_pop = population_map[node.population];
+            }
+            ret = tsk_node_table_add_row(&self->nodes, node.flags, node.time, new_pop,
+                new_ind, node.metadata, node.metadata_length);
+            if (ret < 0) {
+                goto out;
+            }
+            X_node_map[node.id] = ret;
+        }
+    }
+
+    // edges
+    for (k = 0; k < (tsk_id_t) other->edges.num_rows; k++) {
+        tsk_edge_table_get_row(&other->edges, k, &edge);
+        if ((node_mapping[edge.parent] == TSK_NULL)
```

Comment:
here @petrelharp see comment below. we could disable this check if all mutations have non-NaN times.

Reply:
I say disable this check and just add a strict note in the docs saying that if you don't have mutation times and you're using this method then shared history should be ancestral and you better know what you're doing.

Reply:
I'll leave this in here until we get canon sort going and I add it to `tsk_check_subset_equality`.